### PR TITLE
Feature Daily/Hourly Options Support

### LINKED ISSUE #941: Options: Daily backtesting
Per https://github.com/QuantConnect/Lean/issues/939, backtesting options strategies with daily resolution is not currently supported, only minute.

It'd be good if daily was supported.

### LINKED ISSUE #2553: convert options minute data into daily format
Guys, this is my first stab at volunteering to help out here, so go easy on me!  :)   I was recommended to work on converting the options minute data into a daily format.  Problem also described here:  
https://github.com/QuantConnect/Lean/issues/941

#### Expected Behavior
I do need help to make sure I fully understand the requirements.  I think what I need to do is the following:  

Create a standalone application in the Toolbox project that takes all minute data available in a given directory, loop over the entire directory of securities,  for each security: group by day and strike to build day bars from the minute bars.  If I take GOOG as an example: https://github.com/QuantConnect/Lean/tree/master/Data/option/usa/minute/goog

I see there are 3 file types there 

20151224_openinterest_american.zip	
20151224_quote_american.zip	
20151224_trade_american.zip

For quote, I think the the group by and aggregation logic are as follows:  

high - take max of day
low - take low of day
close - last minute bar close
volume - sum volume
open interest - last volume minute bar 

I need help understanding where I can find the layouts for these files, I thought it was in OptionChain.cs but I dont see it there.   Also, the rest of the logic  
 

#### Actual Behavior
This is a new feature 

#### Potential Solution
This is a new feature 

#### Reproducing the Problem
N/A 

#### System Information
N/A

#### Checklist

- [ ] I have completely filled out this template
- [ x ] I have confirmed that this issue exists on the current `master` branch
- [ ] I have confirmed that this is not a duplicate issue by searching [issues](https://github.com/QuantConnect/Lean/issues)

- [ ] I have provided detailed steps to reproduce the issue

### PR DESCRIPTION
**Pull request by C-SELLERS**
#### Description

Lean changes for QuantConnect/LeanCloud.Data#385

- Support daily/hourly options in Lean
- Major refactor LeanDataWriter to streamline writing process
    - Use DataCacheProvider to read and write data
    - Create Generic write that works for all resolutions
    - Create new DiskDataCacheProvider for default use of LeanDataWriter
- Add regressions for Options
- Add necessary data for regression testing
- Add DataCacheProvider tests
- Add additional LeanDataWriter tests

#### Related Issue

Closes #2553
Closes #941

#### How Has This Been Tested?

All tests passing, lots of additional tests added.

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. 
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

File: Common/Interfaces/IDataCacheProvider.cs [hunk 1]
```diff
@@ -43,4 +43,33 @@ public interface IDataCacheProvider : IDisposable
...
+
+    // TODO: Better place for this?
+    public static class DataCacheProviderExtensions {
+        /// <summary>
+        /// Helper to separate filename and entry from a given key
+        /// </summary>
+        /// <param name="key">The key to parse</param>
+        /// <param name="fileName">File name extracted</param>
+        /// <param name="entryName">Entry name extracted</param>
+        public static void ParseKey(string key, out string fileName, out string entryName)
```
Comment (Martin-Molinero) [inline Common/Interfaces/IDataCacheProvider.cs:55]: Suggest moving this to `LeanData` he is the one generating the file and entry names after all

File: Common/Util/LeanDataPathComponents.cs [hunk 2]
```diff
@@ -144,9 +144,10 @@ public static LeanDataPathComponents Parse(string path)
                 ticker = Path.GetFileNameWithoutExtension(path);
                 if (securityType.IsOption())
                 {
-                    // ticker_trade_american
+                    // ticker_year_trade_american
                     var tickerWithoutStyle = ticker.Substring(0, ticker.LastIndexOfInvariant("_"));
-                    ticker = tickerWithoutStyle.Substring(0, tickerWithoutStyle.LastIndexOfInvariant("_"));
+                    var tickerWithoutTickType = tickerWithoutStyle.Substring(0, tickerWithoutStyle.LastIndexOfInvariant("_"));
+                    ticker = tickerWithoutTickType.Substring(0, tickerWithoutTickType.LastIndexOfInvariant("_"));
```
Comment (Martin-Molinero) [inline Common/Util/LeanDataPathComponents.cs:150]: 🤔 why not just a single `IndexOf`?

File: Compression/Compression.cs [hunk 3]
```diff
@@ -135,6 +135,32 @@ public static bool ZipData(string zipPath, IEnumerable<KeyValuePair<string, byte
...
         }
 
+        /// <summary>
+        /// Zips the specified byte array into the zipPath
+        /// </summary>
+        /// <param name="zipPath"></param>
+        /// <param name="zipEntry"></param>
+        /// <param name="data"></param>
+        /// <returns></returns>
+        public static bool ZipData(string zipPath, string zipEntry, byte[] data)
```
Comment (Martin-Molinero) [inline Compression/Compression.cs:145]: hm think this case is handled by `ZipCreateAppendData` already really, suggest not creating a new method if not required

File: Compression/Compression.cs [hunk 4]
```diff
@@ -196,6 +222,37 @@ public static bool ZipCreateAppendData(string path, string entry, string data, b
...
+        /// <param name="data">The entry data</param>
+        /// <param name="overrideEntry">True if should override entry if it already exists</param>
+        /// <returns>True on success</returns>
+        public static bool ZipCreateAppendData(string path, string entry, Byte[] data, bool overrideEntry = false)
+        {
+            try
+            {
+                using (var zip = File.Exists(path) ? ZipFile.Read(path) : new ZipFile(path))
+                {
+                    if (zip.ContainsEntry(entry) && overrideEntry)
```
Comment (Martin-Molinero) [inline Compression/Compression.cs:239]: > zip.ContainsEntry(entry) && overrideEntry

Would be sweeter the other way round 😁

File: Engine/DataFeeds/ZipDataCacheProvider.cs [hunk 5]
```diff
@@ -32,10 +32,10 @@ namespace QuantConnect.Lean.Engine.DataFeeds
     /// </summary>
     public class ZipDataCacheProvider : IDataCacheProvider
     {
-        private const int CacheSeconds = 10;
+        private readonly int _cacheSeconds;
 
         // ZipArchive cache used by the class
-        private readonly ConcurrentDictionary<string, CachedZipFile> _zipFileCache = new ConcurrentDictionary<string, CachedZipFile>();
```
Comment (Martin-Molinero) [inline Engine/DataFeeds/ZipDataCacheProvider.cs:38]: Think can revert this change 👍

File: Engine/DataFeeds/ZipDataCacheProvider.cs [hunk 6]
```diff
@@ -47,55 +47,49 @@ public class ZipDataCacheProvider : IDataCacheProvider
...
             _dataProvider = dataProvider;
-            _cacheCleaner = new Timer(state => CleanCache(), null, TimeSpan.FromSeconds(CacheSeconds), Timeout.InfiniteTimeSpan);
+            _cacheCleaner = new Timer(state => CleanCache(), null, TimeSpan.FromSeconds(_cacheSeconds), Timeout.InfiniteTimeSpan);
         }
 
         /// <summary>
         /// Does not attempt to retrieve any data
         /// </summary>
-        public Stream Fetch(string key)
+        public virtual Stream Fetch(string key)
```
Comment (Martin-Molinero) [inline Engine/DataFeeds/ZipDataCacheProvider.cs:61]: Same, think can revert making fetch virtual?

File: Engine/DataFeeds/ZipDataCacheProvider.cs [hunk 7]
```diff
@@ -307,6 +319,45 @@ private Stream CreateStream(CachedZipFile zipFile, string entryName, string file
...
+                {
+                    cachedZip = new CachedZipFile(dataStream, DateTime.UtcNow, filename);
+
+                    if (!_zipFileCache.TryAdd(filename, cachedZip))
+                    {
+                        // some other thread could of added it already, lets dispose ours
+                        cachedZip.Dispose();
+                    }
+
+                    return true;
```
Comment (Martin-Molinero) [inline Engine/DataFeeds/ZipDataCacheProvider.cs:344]: Hm disposing of the cachedZip but returning true? Think this case should be handled differently  `some other thread could of added it already, lets dispose ours`, should probably fetch the cachedZip in the collection instead?

File: Engine/DataFeeds/ZipDataCacheProvider.cs [hunk 8]
```diff
@@ -132,7 +126,25 @@ public Stream Fetch(string key)
...
+                    // Create the zip, if successful cache it for later use
+                    if (Compression.ZipData(fileName, entryName, data))
+                    {
+                        Cache(fileName, out _);
+                    }
+
+                    return;
+                }
+
+                cachedZip.WriteEntry(entryName, data);
```
Comment (Martin-Molinero) [inline Engine/DataFeeds/ZipDataCacheProvider.cs:146]: hm this method isn't really thread safe, same for `WriteEntry` bellow. Image the cases where the timer kicks in while it's being written too? or if another thread is ready while writing to it?

File: Common/Interfaces/IDataCacheProvider.cs [hunk 9]
```diff
@@ -43,4 +43,33 @@ public interface IDataCacheProvider : IDisposable
         /// <param name="data">The data to cache as a byte array</param>
         void Store(string key, byte[] data);
     }
+
+    // TODO: Better place for this?
```
Comment (C-SELLERS) [inline Common/Interfaces/IDataCacheProvider.cs:47]: Curious if I should move this

File: Engine/DataFeeds/Enumerators/Factories/BaseDataSubscriptionEnumeratorFactory.cs [hunk 10]
```diff
@@ -87,6 +87,12 @@ public IEnumerator<BaseData> CreateEnumerator(SubscriptionRequest request, IData
                     var entriesForDate = factory.Read(source);
                     foreach (var entry in entriesForDate)
                     {
+                        // Fix for Daily/Hour options cases when reading in all equity data from daily/hourly file
```
Comment (C-SELLERS) [inline Engine/DataFeeds/Enumerators/Factories/BaseDataSubscriptionEnumeratorFactory.cs:90]: Sad fix for options enumeration of underlying equity on daily/hourly files

File: Engine/DataFeeds/ZipDataCacheProvider.cs (hunk 6 above)
Comment (C-SELLERS) [inline Engine/DataFeeds/ZipDataCacheProvider.cs:61]: nit - doesn't need to virtual

**Review by C-SELLERS — COMMENTED** [review]
Just a few things I found when doing a self review

File: Tests/ToolBox/LeanDataWriterTests.cs [hunk 11]
```diff
@@ -171,5 +178,195 @@ public void LeanDataWriter_CanWriteCrypto()
...
+                    return DateTime.MinValue;
+                default:
+                    throw new NotImplementedException("This has only implemented a few security types (Equity/Crypto/Option)");
+            }
+        }
+
+        /// <summary>
+        /// Fake brokerage that just uses Local Disk Data to do history requests
+        /// </summary>
+        public class LocalHistoryBrokerage : NullBrokerage
```
Comment (Martin-Molinero) [inline Tests/ToolBox/LeanDataWriterTests.cs:315]: > public class LocalHistoryBrokerage : NullBrokerage 

Cool! Can make these internal 👍

File: Tests/ToolBox/LeanDataWriterTests.cs [hunk 12]
```diff
@@ -171,5 +178,195 @@ public void LeanDataWriter_CanWriteCrypto()
...
+            private readonly DefaultDataProvider _dataProvider;
+            private IDataCacheProvider _dataCacheProvider;
+            private IHistoryProvider _historyProvider;
+
+            public LocalHistoryBrokerage()
+            {
+                var mapFileProvider = new LocalDiskMapFileProvider();
+                _dataProvider = new DefaultDataProvider();
+                _dataCacheProvider = new ZipDataCacheProvider(_dataProvider);
+                var factorFileProvider = new LocalDiskFactorFileProvider();
```
Comment (Martin-Molinero) [inline Tests/ToolBox/LeanDataWriterTests.cs:326]: Should use `TestGlobals.` for the factor/map/data providers

File: Common/Data/LeanDataWriter.cs [hunk 13]
```diff
@@ -172,340 +214,178 @@ public void DownloadAndSave(IBrokerage brokerage, List<Symbol> symbols, DateTime
...
+            else
             {
-                WriteFile(outputFile, lines, lastTime);
+                // Write out this data string to a zip file
+                var tempFilePath = filePath + ".tmp";
+                Compression.ZipData(tempFilePath, entryName, rows.Values);
+
+                // Move temp file to the final destination with the appropriate name
+                File.Move(tempFilePath, filePath);
+                Log.Trace("LeanDataWriter.Write(): Created: " + filePath);
```
Comment (Martin-Molinero) [inline Common/Data/LeanDataWriter.cs:314]: Think ideally should also use the `_dataCacheProvider.Store` here? Think we already handled the `_appendToZips` behavior deleting the file if it existed?

File: Common/Data/LeanDataWriter.cs [hunk 14]
```diff
@@ -172,340 +214,178 @@ public void DownloadAndSave(IBrokerage brokerage, List<Symbol> symbols, DateTime
...
-                            rows[time] = line;
-                        }
-                    }
-                }
-            }
+    /// <summary>
+    /// Simple data cache provider, writes and reads directly from disk
+    /// Used as default for <see cref="LeanDataWriter"/>
+    /// </summary>
+    public class DiskDataCacheProvider : IDataCacheProvider
```
Comment (Martin-Molinero) [inline Common/Data/LeanDataWriter.cs:335]: Cool! let's move this guy to his own class file plus  a couple of tests would be great! 👍

**Review by Martin-Molinero — COMMENTED** [review]
Sweet improvement! `LeanDataWriter` is *much* better
Leaving a few comments and request. I think the changes in the ZipDataCacheProvider need a new set of unit tests.
Think we need to add an options mapping regression test, believe the behavior is cover already but would be nice to assert it, for example we have the `OptionRenameRegressionAlgorithm` for minute resolution

File: Common/Data/LeanDataWriter.cs (hunk 13 above)
Comment (C-SELLERS) [reply Common/Data/LeanDataWriter.cs:314]: I think this is unique here, in that it would replace the entire zip. Not just the entry in the zip. Not sure where or why we would use it this way 🤔

Comment (C-SELLERS) [reply Common/Data/LeanDataWriter.cs]: Solved 👍🏽  _appendToZip is not necessary. 

There is no case in which we will want to delete the zip rather than just replace the entry. By replacing the entry we are deleting the entry itself and writing a new one in its place. Leave the zip alone.

This also address issues with the DataCacheProvider having the file open when this part would attempt to delete. 

End result is to let the DataCacheProvider handle the writing process entirely, do not attempt to modify files in LeanDataWriter.

File: Tests/ToolBox/LeanDataWriterTests.cs [hunk 15]
```diff
@@ -171,5 +178,193 @@ public void LeanDataWriter_CanWriteCrypto()
 
             Assert.AreEqual(data.First().Value.Count(), 3);
         }
+
+        [Explicit("Data copying and reading test")]
```
Comment (Martin-Molinero) [inline Tests/ToolBox/LeanDataWriterTests.cs:182]: By using `Explicit` these test aren't being run by the CI, think we should remove it? `CanDownloadAndSave(Equity,Quote,Hour)` is throwing some exceptions, my guess is because the file doesn't exist?

File: Tests/ToolBox/LeanDataWriterTests.cs [hunk 16]
```diff
@@ -171,5 +178,193 @@ public void LeanDataWriter_CanWriteCrypto()
...
+        {
+            private readonly IDataCacheProvider _dataCacheProvider;
+            private readonly IHistoryProvider _historyProvider;
+
+            public LocalHistoryBrokerage()
+            {
+                var mapFileProvider = TestGlobals.MapFileProvider;
+                var dataProvider = TestGlobals.DataProvider;
+                _dataCacheProvider = new ZipDataCacheProvider(dataProvider);
+                var factorFileProvider = new LocalDiskFactorFileProvider();
```
Comment (Martin-Molinero) [inline Tests/ToolBox/LeanDataWriterTests.cs:325]: There is a `TestGlobals.FactorFileProvider` available to use too 👍

File: Common/Data/LeanDataWriter.cs [hunk 17]
```diff
@@ -172,328 +213,99 @@ public void DownloadAndSave(IBrokerage brokerage, List<Symbol> symbols, DateTime
...
+
+            // Handle merging of files
+            // Only merge on files with hour/daily resolution, that exist, and can be loaded
+            string finalData;
+            if (_resolution >= Resolution.Hour && fileExists && TryLoadFile(filePath, entryName, out var rows))
             {
-                // Ensure the data is sorted
-                if (data.Time < lastTime) throw new Exception("The data must be pre-sorted from oldest to newest");
+                // Load this data into a sorted dictionary so we may commence the merge
+                var sortedNewData = data.ToImmutableSortedDictionary(x => x.Item1, x => x.Item2);
```
Comment (Martin-Molinero) [inline Common/Data/LeanDataWriter.cs:282]: Not sure `ToImmutableSortedDictionary` is required, can directly iterate over `data` bellow?

**Review by Martin-Molinero — APPROVED** [review]
Awesome minor final comments ✨

File: Tests/ToolBox/LeanDataWriterTests.cs (hunk 15 above)
Comment (C-SELLERS) [reply Tests/ToolBox/LeanDataWriterTests.cs:182]: Gotcha 👌🏽 Removed explicit tag.

After reviewing `CanDownloadAndSave(Equity,Quote,Hour)`, yes you are correct, it was reading in TradeBars from our Hourly folder as QuoteBars.

File: Common/Data/LeanDataWriter.cs (hunk 17 above)
Comment (C-SELLERS) [reply Common/Data/LeanDataWriter.cs:282]: Oh nice! Yes that would work, I'm glad too because I really didn't like how it was.

**Review by Martin-Molinero — APPROVED** [review]
Sweet! CI is failing in a few tests though
